### PR TITLE
nixos/luksroot: GPG Smartcard support for luks encrypted volumes in initrd

### PR DESCRIPTION
**Pull request by marenz2569**
###### Motivation for this change

I do want to decrypt my luks volumes with a gpg smart card.
You manually have to encrypt the passphrase for the decryption of luks and specify the gpg encrypted file.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Assured whether relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: nixos/modules/system/boot/luksroot.nix [hunk 1]
```diff
@@ -470,6 +571,37 @@ in
             '';
           };
 
+          gpgCard = mkOption {
+            default = null;
+            description = ''
+              The option to use this LUKS device with a GPG encrypted luks password by the GPG Smartcard.
+              If null (the default), GPG-Smartcard will be disabled for this device.
+              Yubikey and GPG Card may not be used at the same time.
```
Comment (artemist) [inline nixos/modules/system/boot/luksroot.nix]: If you can't use yubikey and gpg card at the same time, you may want to add an assertion to `config.assertions`.

Comment (marenz2569) [reply nixos/modules/system/boot/luksroot.nix]: @artemist I have added the assertion.

**Comment by petrkr** [comment]
This PR broken older PR for trezor support, check #60774

File: nixos/modules/system/boot/luksroot.nix [hunk 2]
```diff
@@ -275,16 +309,83 @@ let
...
+        if wait_gpgcard ${toString gpgCard.gracePeriod}; then
+            do_open_gpg_card
+        else
+            echo "No GPG Card found, falling back to normal open procedure"
+            open_normally
+        fi
+    }
+    ''}
+
+    ${if (luks.yubikeySupport && (yubikey != null)) || (luks.gpgSupport && (gpgCard != null)) then ''
```
Comment (petrkr) [inline nixos/modules/system/boot/luksroot.nix]: So if there will be 10 or 20 HW ways how to open wallet, this IF will be endless ?

**Pull request merged**